testtools/UART_interface/rpi_uart_interface.py

```diff
@@ -0,0 +1,197 @@
+
+
+def check_firmware_errors(line):
+    if azure_test_firmware_errors.iot_init_failure in line:
+        print("Failed to connect to saved IoT Hub!")
+        azure_test_firmware_errors.SDK_ERRORS += 1
+
+    elif azure_test_firmware_errors.sensor_init_failure in line:
+        print("Failed to init mxchip sensor")
+        azure_test_firmware_errors.SDK_ERRORS += 1
+
+    elif azure_test_firmware_errors.wifi_failure in line:
+        print("Failed to connect to saved WiFi network.")
+        azure_test_firmware_errors.SDK_ERRORS += 1
+
+# ------- interface class -------
+class rpi_uart_interface(uart_interface):
+
+    # If there is a sudden disconnect, program should report line in input script reached, and close files.
+    # method to write to serial line with connection monitoring
+    def serial_write(self, ser, message, file=None):
+
+        # Check that the device is no longer sending bytes
+        if ser.in_waiting:
+            self.serial_read(ser, message, file)
+
+        # Check that the serial connection is open
+        if ser.writable():
+
+            # special handling for sending a file
+            if "sz " in message:
+                round = ser.write(bytearray("rz\r\n".encode('ascii'))) #
+                os.system(message)#"sz -a test.sh > /dev/ttyUSB0 < /dev/ttyUSB0")
+
+                # check for completion
+                output = ser.readline(ser.in_waiting)
+                output = output.decode(encoding='utf-8', errors='ignore')
+                # print(output)
+                start_time = time.time()
+                while ("Transfer" not in output) and (time.time() - start_time) < serial_settings.wait_for_flash:
+                    time.sleep(.01)
+                    output = ser.readline(ser.in_waiting)
+                    output = output.decode(encoding='utf-8', errors='ignore')
+
+                return round
+            # special handling for receiving a file, Note: input file should use rz <filepath> when a file from the RPi is desired
+            elif "rz " in message:
+                os.system('rz')
+                message.replace('rz ', 'sz ')
+
+            # wait = serial_settings.mxchip_buf_pause # wait for at least 50ms between 128 byte writes. -- may not be necessary on rpi
+            buf = bytearray((message).encode('ascii'))
+            print(buf)
+            # buf_len = len(buf)  # needed for loop as buf is a destructed list
+            # bytes_written = 0
+            bytes_written = ser.write(buf)
+            # timeout = time.time()
+            # while bytes_written < buf_len:
+            #     round = ser.write(buf[:128])
+            #     buf = buf[round:]
+            #     bytes_written += round
+            #     # print("bytes written: %d" %bytes_written)
+            #     time.sleep(wait)
+            #     if (time.time() - timeout > serial_settings.serial_comm_timeout):
+            #         break
+            # print("final written: %d" %bytes_written)
```

Comment:
nit: Remove comments if not needed

Reply:
Throughout this file and other files actually. If it helps debugging, resolve this and remove in next round. 

---
In reply to: [295057201](https://github.com/Azure/azure-iot-sdk-c/pull/1051#discussion_r295057201) [](ancestors = 295057201)